### ISSUE TITLE
NativeBytesStore.toTemporaryDirectByteBuffer() doesn't strongly reference the memory it points to

### ISSUE THREAD

**Issue by aweisberg** [OpenHFT/Chronicle-Bytes#69]
I was having issues where deserialization would fail with an out of range value here https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/tools/fqltool/Dump.java#L84

If I change it toByteArray() it works consistently. java.nio.DirectByteBuffer has an attachment field you can set to whatever it needs to strongly reference.

I can't say with 100% certainty this is what is happening, since it is surprising to get so unlucky. It fails the first time almost every time.

**Comment by dpisklov** [comment]
From quick glance at what you are doing, I wouldn't recommend using `#toTemporaryDirectByteBuffer()`. You can use `#underlyingObject()` instead, and the underlying object will be ByteBuffer.
In addition, `codec#decode()` could directly use net.openhft.chronicle.bytes.Bytes, instead of ByteBuffer...

**Comment by dpisklov** [comment]
I'm also not sure `#toTemporaryDirectByteBuffer()` is doing what you expect it to do:
```
default ByteBuffer toTemporaryDirectByteBuffer() throws IllegalArgumentException {
        int len = Maths.toUInt31(readRemaining());
        try {
            ByteBuffer bb = ByteBuffer.allocateDirect(len);
            copyTo(bb);
            bb.clear();
            return bb;
        } catch (BufferUnderflowException e) {
            throw new AssertionError(e);
        }
    }
```
Note `bb.clear()` - it clears the ByteBuffer hence your error...

**Comment by aweisberg** [comment]
What's wrong with bb.clear()? https://docs.oracle.com/javase/7/docs/api/java/nio/Buffer.html#clear()
It's not the idiomatic thing to use (that would be flip()) but it works. It doesn't actually clear the contents of the buffer.

toTemporaryDirectByteBuffer() might not be the best method, but as long as it's there it should probably not do something surprising like that. If it is going to do it then it should probably have some javadoc warning that it's not going to strongly reference the memory it points to.

**Comment by dpisklov** [comment]
Well clear() resets the pointers so without first setting position/limit to where it was unlikely you will be able to read the data. I think this is what causing your issue, and not the lack of strong reference. While you're inside the reading document (and your whole reader closure is inside) the memory won't be free'd.

**Comment by aweisberg** [comment]
It's a copy. The only bytes in there are the bytes I care about right? So clear just sets it to the beginning of the bytes I care about and limit set to capacity is the end of the bytes I care about.

**Comment by dpisklov** [comment]
It sounds reasonable to me after looking at the ByteBuffer code but it's somewhat counter-intuitive...
Anyway given we actually do copy bytes into that ByteBuffer I don't think it makes sense to reference some other memory from it? Or am I missing something?

**Comment by aweisberg** [comment]
Ah that is the wrong implementation of toTemporaryDirectByteBuffer(). Sorry didn't notice.

https://github.com/OpenHFT/Chronicle-Bytes/blob/cbb5de9d77bdae82789dad141266a8b28a8214b7/src/main/java/net/openhft/chronicle/bytes/NativeBytesStore.java#L688

```
  public ByteBuffer toTemporaryDirectByteBuffer() {
        ByteBuffer bb = ByteBuffer.allocateDirect(0);
        try {
            BB_ADDRESS.setLong(bb, address);
            BB_CAPACITY.setInt(bb, Maths.toUInt31(readRemaining()));

        } catch (Exception e) {
            throw new AssertionError(e);
        }
        bb.clear();
        return bb;
    }
```

**Comment by dpisklov** [comment]
Actually I just realised that you're doing `WireIn#bytesStore()` - any reason for this? This creares a new BytesStore which wraps bytes and does all sort of copying etc (and which is then I guess discarded by JVM as it probably is optimised and allocated on stack...).
You can (and probably should) instead use either WireIn#bytes() which will give you byte[] if that suits you, or alternatively you can use WireIn#bytes(Bytes toBytes), and give it reusable Bytes object (which can be a simple wrapper over ByteBuffer).

**Comment by aweisberg** [comment]
[Switching to bytes() is the plan.](https://github.com/apache/cassandra/compare/trunk...aweisberg:cassandra-14504-trunk?expand=1#diff-6665086f6cb49a99e69936c0c3ed8b36L84)

I think I incorrectly assumed bytesStore() would return a zero copy view of the underlying storage and that would be a way for me to get a ByteBuffer pointing to the mmapped file so I could construct the Netty ByteBuf necessary for deserialization.

Making a copy is fine as this isn't the most performance sensitive use case.

**Comment by dpisklov** [comment]
OK so you mind closing this issue then if your problem is solved?
Thanks

**Comment by aweisberg** [comment]
This issue isn't about a problem in Cassandra it's about NativeBytesStore.toTemporaryDirectByteBuffer()? I know we got sidetracked, but it really seems like that needs to be either implemented or documented better.

**Comment by dpisklov** [comment]
OK I'll think of a way to improve the documentation

**Comment by aweisberg** [comment]
Is there a reason not set the DirectByteBuffer.att to reference the memory it points to? Then it would just work even for people who don't read the javadoc instead of failing at runtime non-deterministically.

**Comment by dpisklov** [comment]
The problem is there's nothing to point to. We allocate memory using `Unsafe.allocateMemory()` and the only thing we have is address. Then BytesStore keeps track of the references using reserve/release and when last release is called the ref cleaner deallocates the memory. The problem is if we try to reserve when we do `toTemporaryDirectBuffer()` then we somehow need to be sure release() will be called and that's impossible to guarantee in this scenario, hence we shouldn't do that. So the only possibility to me is to update doco to specify this behaviour.

**Comment by aweisberg** [comment]
I think you are right there are cases where explicit memory management is used, but those tend to work correctly as long as you don't leak the reference to temporary DBB beyond the scope where the reference is released. I'm not doing that in this case. 

I think there is a bug here it's just not the one I thought.

So BinaryWire.bytesStore() in the u8_array and i64_array case invokes NativeBytesStore.lazyNativeBytesStoreWithFixedCapacity() and makes a copy. This memory that is copied doesn't use explicit memory management. You can see that in NativeBytesStore.of() where it creates a deallocator and then a cleaner for the deallocator. So it is using GC and strong references so if the DBB that toTemporaryDirectBuffer() creates references the NativeBytesStore it points to then that won't be reclaimed.

If explicit memory management were in use I should have been OK because I would have finished my processing before release is invoked. I'm certainly not invoking release and I don't see an opportunity for that to occur in the code I linked to until after deserialization has occurred.

**Comment by dpisklov** [comment]
@aweisberg sorry it took me too long to get back on this.
I took your suggestion and now we store strong reference. Thanks

**Comment by hft-team-city** [comment]
Released in [Chronicle-Bytes-2.20.101](https://github.com/OpenHFT/Chronicle-Bytes/releases/chronicle-bytes-2.20.101), [BOM-2.20.134](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.20.134)
